docs/decisions/0011-producing-to-event-bus-via-settings.rst

```diff
@@ -0,0 +1,92 @@
+   # .. setting_description: Dictionary specifying event_types (keys) and its configuration dictionary as values.
+   #   Each configuration dictionary contains a flag denoting whether the event will be enabled or disabled and
+   #   optionally details like topics/stream names where the event will be pushed to, event_key_field which is
+   #   and period delimited string path to event data field to use as event key.
+   #   Note: The topic names should not include environment prefix as it will be dynamically added based on
+   #   EVENT_BUS_TOPIC_PREFIX setting.
+   EVENT_BUS_PRODUCER_CONFIG = {
+       'org.openedx.content_authoring.xblock.deleted.v1': {
+           'event_key_field': 'xblock_info.usage_key', # override default event_key_field
+           'enabled': True,
+       },
+       'org.openedx.content_authoring.xblock.published.v1': {
+           'topics': ['content-authoring-xblock-lifecycle', 'xblock-published'], # override default topics
+           'enabled': True,
+       },
+       'org.openedx.content_authoring.xblock.duplicated.v1': {
+           'enabled': True, # enable this event and use default values
+       },
+   }
+
+This configuration will be read and combined with the default config in openedx_events ``apps.OpenedxEventsConfig(AppConfig).ready`` method and a generic signal handler will be connected to the event_types (keys) listed in the configuration after validating its format.
+
+.. code-block:: python
+
+   def ready(self):
+       load_all_signals()
+       config = read_config()
+       # validate the config dictionary and raise errors if any
+       validate_config()
+       for signal in all_signals:
+           if (topic and event_key_field is configured) and enabled:
+               connect_or_disconnect_handlers(configured_signal)
+
+The generic handler will again read the configuration and get details for event or signal triggered and push it to the event bus.
+
+.. code-block:: python
+
+   def general_signal_handler(sender, signal, **kwargs):
+       config = read_config(signal)
+       for topic in topics:
+           push_event_to_implementation(signal, config)
+
+
+Optionally, add a function to ``OpenEdxPublicSignal`` to return default configuration for documentating them.
+
+Consequences
+************
+
+* Applications producing to event_bus can push events without requiring any additional code.
+* Users can push multiple related event_types to the same topic, which will allow them to run a single consumer and process the events in the correct order.
+* The configuration is a dictionary, which makes it flexible but difficult to enforce schemas and maintain them.
+* Producing applications similar to consuming applications will be required to add ``openedx_events`` as an application in their ``INSTALLED_APPS`` settings.
+* Adding details like ``topic`` and ``event_key_field`` to signal definition will keep configuration compact if user does not want to override them.
```

Comment:
>I could maybe see event_key_field being part of the definition

Agreed, we can maybe keep `event_key_field` with the definition and remove `topics`. This can be defined via the configuration only and add defaults to each service as you suggested above.

Reply:
I think I'd prefer to keep both topic and event key with the service config. Centralizing the information to the event definition could allow us to avoid repetition in configs, but I don't think any event is going to be produced by more than one service in a deployment; in general, events should be produced by one service and then consumed by one or more others. The configs also don't have any environment-specific information in them, so we shouldn't even see repetition between stage vs. production configs—this config would be able to live in the service's own base settings.

Given the expected lack of repetition, I think having the topic and key field be closer to the point of production rather than farther (requiring library releases in order to change) would make config changes easy, without even the need for a default/override system.

Reply:
> events should be produced by one service and then consumed by one or more others.

@timmc-edx Makes absolute sense. Since we won't have any repetition as it is, we don't need to add these to the signal definition and complicate it. I'll update the ADR and mention that we will include the default config in all the producer applications as a starting point.

Reply:
Updated.

Reply:
Sounds good to me.